### PR TITLE
feat: changes for supporting record event in FB audience

### PR DESCRIPTION
**Pull request by Vikas26021999**
## What are the changes introduced in this PR?

changes for supporting record event in FB audience

## What is the related Linear task?

Resolves INT-XXX

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3351/test-report.html

**Review by yashasvibajpai — COMMENTED** [review]
Lets add tests too. :) 

1. SImple insert record
2. Simple delete record
3. Simple update record
4. Batching multiple record events with a custom batch size config
5. Failure scenarios like instrumentation errors etc

File: src/v0/destinations/fb_custom_audience/transformV2.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  }
+  if (type && type !== 'NA' && typeFields.includes(type)) {
+    dataSource.type = type;
+  }
+  if (subType && subType !== 'NA' && subTypeFields.includes(subType)) {
+    dataSource.sub_type = subType;
+  }
+  if (Object.keys(dataSource).length > 0) {
+    paramsPayload.data_source = dataSource;
+  }
```
Comment (ItsSudip) [inline src/v0/destinations/fb_custom_audience/transformV2.js:169]: This is repeated with audienceList type implementation can't we move this to utils and use it in both places?

File: src/v0/destinations/fb_custom_audience/transform.js [hunk 2]
```diff
@@ -258,7 +226,24 @@ const processEvent = (message, destination) => {
...
+  }
+
+  if (groupedInputs.audiencelist) {
+    transformedAudienceEvent = await simpleProcessRouterDest(
+      groupedInputs.audiencelist,
+      process,
+      reqMetadata,
+    );
+  }
+
```
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transform.js:245]: should we ignore them? or should we throw an error here?

File: src/v0/destinations/fb_custom_audience/transformV2.js [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  };
+
+  // maxUserCount validation
+  const maxUserCountNumber = parseInt(maxUserCount, 10);
+  if (Number.isNaN(maxUserCountNumber)) {
+    throw new ConfigurationError('Batch size must be an Integer.');
+  }
+
+  // audience id validation
+  const { audienceId } = destination.Config;
```
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transformV2.js:133]: done

File: src/v0/destinations/fb_custom_audience/transformV2.js (hunk 1 above)
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transformV2.js:169]: updated

File: src/v0/destinations/fb_custom_audience/transform.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transform.js:245]: just need to throw error in this, confirmed offline on a call

File: src/v0/destinations/fb_custom_audience/transformV2.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  let operationAudienceId = audienceId;
+  const mappedToDestination = get(message, MappedToDestinationKey);
+  if (!operationAudienceId && mappedToDestination) {
+    const { objectType } = getDestinationExternalIDInfoForRetl(message, 'FB_CUSTOM_AUDIENCE');
+    operationAudienceId = objectType;
+  }
```
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transformV2.js:181]: discussed this offline

**Review by yashasvibajpai — APPROVED** [review]

File: test/integrations/destinations/fb_custom_audience/router/data.ts
Comment (ItsSudip) [inline test/integrations/destinations/fb_custom_audience/router/data.ts:1]: Can we update the test case to verify the `maxUserCount`? I can see you have set it to 3, now to verify that we should have more than 3 user details in any batches.

Comment (Vikas26021999) [reply test/integrations/destinations/fb_custom_audience/router/data.ts:1]: updated

**Review by yashasvibajpai — APPROVED** [review]

**Review by ItsSudip — DISMISSED** [review]

File: src/v0/destinations/fb_custom_audience/transformV2.js [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+  if (finalResponse.length === 0) {
+    throw new InstrumentationError(
+      'Missing valid parameters, unable to generate transformed payload',
+    );
+  }
+  return finalResponse;
+}
+
+module.exports = {
+  processRecordInputs,
```
Comment (krishna2020) [inline src/v0/destinations/fb_custom_audience/transformV2.js:309]: Move responseBuilderSimple & getDataSource to util class as they are common in both audienceList & recordType flow
Rename `transformV2.js` to `recordTransform.js` as v2 class is not a complete parallel path to transform.js

Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transformV2.js]: addressed

File: src/v0/destinations/fb_custom_audience/util.js [hunk 6]
```diff
@@ -221,9 +227,44 @@ const generateAppSecretProof = (accessToken, appSecret, dateNow) => {
...
+    return response;
+  }
+  // fail-safety for developer error
+  throw new TransformationError(`Payload could not be constructed`);
+};
+
 module.exports = {
   prepareDataField,
   getSchemaForEventMappedToDest,
   batchingWithPayloadSize,
```
Comment (krishna2020) [inline src/v0/destinations/fb_custom_audience/util.js:264]: @ItsSudip Were these checks ensureApplicableFormat & getUpdatedDataElement missing in the old audienceList flow ?

File: src/v0/destinations/fb_custom_audience/recordTransform.js [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        });
+      }
+      data.push(dataElement);
+      metadata.push(input.metadata);
+    });
+
+    const prepareFinalPayload = lodash.cloneDeep(paramsPayload);
+    prepareFinalPayload.schema = userSchema;
+    prepareFinalPayload.data = data;
+    const payloadBatches = batchingWithPayloadSize(prepareFinalPayload);
```
Comment (krishna2020) [inline src/v0/destinations/fb_custom_audience/recordTransform.js:95]: Is payload size limitation in batch a new requirement from FB ?

File: src/v0/destinations/fb_custom_audience/util.js
Comment (krishna2020) [inline src/v0/destinations/fb_custom_audience/util.js:1]: Need unit tests for `util.js` method, can add `utils.test.js` in the same directory - can refer https://github.com/rudderlabs/rudder-transformer/blob/main/src/v0/destinations/ga4/utils.test.js

File: test/integrations/destinations/fb_custom_audience/router/data.ts
Comment (krishna2020) [inline test/integrations/destinations/fb_custom_audience/router/data.ts:1]: this file is impossible to read - reached 50k lines
we are following new pattern to make them more readable by dividing test cases to different files based on scenarios and initialise common information like config at global level & re-use them across test cases for reducing length of test case.
https://github.com/rudderlabs/rudder-transformer/tree/develop/test/integrations/destinations/klaviyo  --> you can check Klaviyo one.

File: src/v0/destinations/fb_custom_audience/recordTransform.js [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+  // maxUserCount validation
+  const maxUserCountNumber = parseInt(maxUserCount, 10);
+  if (Number.isNaN(maxUserCountNumber)) {
+    throw new ConfigurationError('Batch size must be an Integer.');
+  }
+
+  // audience id validation
+  let operationAudienceId = audienceId;
+  const mappedToDestination = get(message, MappedToDestinationKey);
+  if (!operationAudienceId && mappedToDestination) {
```
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/recordTransform.js:145]: updated

File: src/v0/destinations/fb_custom_audience/transform.js [hunk 9]
```diff
@@ -265,7 +238,29 @@ const processEvent = (message, destination) => {
...
 const processRouterDest = async (inputs, reqMetadata) => {
-  const respList = await simpleProcessRouterDest(inputs, process, reqMetadata);
+  const respList = [];
+  const groupedInputs = lodash.groupBy(inputs, (input) => input.message.type?.toLowerCase());
+  let transformedRecordEvent = [];
+  let transformedAudienceEvent = [];
+
+  const eventTypes = ['record', 'audiencelist'];
+  if (extraKeysPresent(groupedInputs, eventTypes)) {
+    throw new ConfigurationError('unsupported events present in the event');
```
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/transform.js:248]: updated

File: src/v0/destinations/fb_custom_audience/recordTransform.js [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    operationAudienceId = objectType;
+  }
+  if (!isDefinedAndNotNullAndNotEmpty(operationAudienceId)) {
+    throw new ConfigurationError('Audience ID is a mandatory field');
+  }
+
+  // user schema validation
+  let { userSchema } = destination.Config;
+  if (mappedToDestination) {
+    userSchema = getSchemaForEventMappedToDest(message);
```
[the remark this answers is not in the document]
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/recordTransform.js:156]: yes

File: src/v0/destinations/fb_custom_audience/util.js (hunk 6 above)
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/util.js:264]: @krishna2020 these are also present in audience flow, they are called with `prepareDataField` in transform.js

File: src/v0/destinations/fb_custom_audience/recordTransform.js (hunk 7 above)
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/recordTransform.js:95]: this was present in audience flow as well

File: src/v0/destinations/fb_custom_audience/util.js
Comment (Vikas26021999) [reply src/v0/destinations/fb_custom_audience/util.js:1]: added tests for important methods, will take up a backlog item for testing remaining functions

File: test/integrations/destinations/fb_custom_audience/router/data.ts
Comment (Vikas26021999) [reply test/integrations/destinations/fb_custom_audience/router/data.ts:1]: refactored this

Comment (krishna2020) [reply test/integrations/destinations/fb_custom_audience/router/data.ts:1]: One more thing pending is on the test structure refactoring,
Prepare two different files - one for `audienceList.ts` & other for `record.ts` & mention these two files under `data.ts`
Few more keys are added in the test structure like successCriteria, scenario - you can check here [https://github.com/rudderlabs/rudder-transformer/blob/457a18b2aec03aa0dfafcadc611b[…]est/integrations/destinations/klaviyo/processor/ecomTestData.ts](https://github.com/rudderlabs/rudder-transformer/blob/457a18b2aec03aa0dfafcadc611b5f7176e97beb/test/integrations/destinations/klaviyo/processor/ecomTestData.ts#L42)
And for metadata object, use generateMetadata(<jobId>) function

Comment (Vikas26021999) [reply test/integrations/destinations/fb_custom_audience/router/data.ts:1]: updated

File: src/types/index.ts [hunk 11]
```diff
@@ -131,6 +131,8 @@ type Destination = {
   WorkspaceID: string;
   Transformations: UserTransformationInput[];
   RevisionID?: string;
```
Comment (krishna2020) [inline src/types/index.ts:133]: Is rudder-server sending these 2 keys to rudder-transformer as of today in production ?

**Review by krishna2020 — APPROVED** [review]

File: src/types/index.ts (hunk 11 above)
Comment (Vikas26021999) [reply src/types/index.ts:133]: yes, confirmed this with @yashasvibajpai , but not being used  in transformer

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**